src/exe_reader.cpp

```diff
@@ -0,0 +1,189 @@
+	// 10
+	exfile.put((hdrL) & 0xFF);
+	exfile.put((hdrL >> 8) & 0xFF);
+	exfile.put((hdrL >> 16) & 0xFF);
+	exfile.put((hdrL >> 24) & 0xFF);
+
+	corefile.seekg(pos, std::ios_base::beg);
+	while (len > 0) {
+		int v = corefile.get();
+		if (v == -1)
+			break;
+		exfile.put(v);
+		len--;
+	}
+	
+	exfile.close();
+}
+
+static uint32_t exe_reader_roffset(uint32_t bas, uint32_t ofs) {
+	return bas + (ofs ^ 0x80000000);
+}
+
+void EXEReader::GetExfont(std::string file) {
+	// Part 2 of the resource grabber.
+	if (!resource_ofs) {
+		Output::Debug("EXEReader::GetExfont : No resource section.");
+		return;
+	}
+	// For each ID/Name entry in the outer...
+	uint32_t resourcesIDEs = GetU16(resource_ofs + 0x0C) + (uint32_t) GetU16(resource_ofs + 0x0E);
+	uint32_t resourcesIDEbase = resource_ofs + 0x10;
+	while (resourcesIDEs) {
+		// This can only be 2 if valid. Don't worry about it.
+		if (GetU32(resourcesIDEbase) == 2) {
+			uint32_t bitmapDBase = exe_reader_roffset(resource_ofs, GetU32(resourcesIDEbase + 4));
+			// Looking for a named entry.
+			uint16_t resourcesNDEs = GetU16(bitmapDBase + 0x0C) + (uint32_t) GetU16(bitmapDBase + 0x0E);
+			uint32_t resourcesNDEbase = bitmapDBase + 0x10;
+			while (resourcesNDEs) {
+				uint32_t name = GetU32(resourcesNDEbase);
+				// Actually a name?
+				if (name & 0x80000000) {
+					name = exe_reader_roffset(resource_ofs, name);
+					// Output::Debug("EXEReader::GetExfont : Name at %x.", name);
+					if (ResNameCheck(name, "EXFONT")) {
+						uint32_t dataent = GetU32(resourcesNDEbase + 4);
+						if (dataent & 0x80000000) {
+							dataent = exe_reader_roffset(resource_ofs, dataent);
+							dataent = resource_ofs + GetU32(dataent + 0x14);
+						}
+						uint32_t filebase = (GetU32(dataent) - resource_rva) + resource_ofs;
+						uint32_t filesize = GetU32(dataent + 0x04);
+						Output::Debug("EXEReader::GetExfont : EXFONT found (DE %x ; %x ; len %x), committing to extraction.", dataent, filebase, filesize);
+						exe_reader_perform_exfont_save(file, corefile, filebase, filesize);
+						return;
+					}
+				}
+				resourcesNDEbase += 8;
+				resourcesNDEs--;
+			}
+			Output::Debug("EXEReader::GetExfont : EXFONT not found in dbase at %x.", bitmapDBase);
+			return;
+		}
+		resourcesIDEbase += 8;
+		resourcesIDEs--;
+	}
+	Output::Debug("EXEReader::GetExfont : BITMAP not found.");
+	return;
+}
+
+uint8_t EXEReader::GetU8(uint32_t i) {
+	corefile.seekg(i, std::ios_base::beg);
+	int ch = corefile.get();
+	if (ch == -1)
+		ch = 0;
+	return (uint8_t) ch;
+}
+
+uint16_t EXEReader::GetU16(uint32_t i) {
+	uint16_t v = GetU8(i);
+	v |= ((uint32_t) GetU8(i + 1)) << 8;
+	return v;
+}
+
+uint32_t EXEReader::GetU32(uint32_t i) {
+	uint32_t v = GetU16(i);
+	v |= ((uint32_t) GetU16(i + 2)) << 16;
+	return v;
+}
+
+bool EXEReader::ResNameCheck(uint32_t i, const char * p) {
+	if (GetU16(i) != strlen(p))
+		return false;
+	while (*p) {
+		i += 2;
+		if (GetU16(i) != *p)
+			return false;
+		p++;
+	}
+	return true;
```

Comment:
Does the code support big endian?

Reply:
Big-endian platforms or big-endian EXEs?

Reply:
Platforms of course, there is no RPG_RT for PowerPC (luckily)

Reply:
Yes. See GetU32/GetU16/GetU8. For simplicity's sake, GetU32 is based off of GetU16 which is based off of GetU8. Thus, no endian-specific code to worry about.